blog/2023-03-28-how-to-use-witnet-oracle-to-display-cryptocurrency-prices/2023-03-28-how-to-use-witnet-oracle-to-display-cryptocurrency-prices.md

```diff
@@ -0,0 +1,277 @@
+authors:
+  - name: Justin Obi
+    title: Software Engineer / Technical Writer
+    url: https://github.com/cjustinobi
+    image_url: https://github.com/cjustinobi.png
+tags: [celosage, intermediate, react, hardhat, oracle]
+hide_table_of_contents: false
+slug: /tutorials/how-to-use-witnet-oracle-to-display-cryptocurrency-prices
+---
+
+![header](../../src/data-tutorials/showcase/intermediate/how-to-use-witnet-oracle-to-display-cryptocurrency-prices.png)
+
+## Introduction​
+
+There are a couple of oracle platforms one could use to get off-chain data to one's Dapp. 
+I want to show how to use Witnet to display price feeds of the popular blockchain networks.
+
+## Prerequisites​
+
+To successfully follow along in this tutorial, you need basic knowledge of the following:
+HTML, CSS, and React
+Blockchain, Solidity, and Hardhat
+
+## Requirements​
+
+- [Remix](https://remix.ethereum.org/) - Online editor for writing smart contracts or any code editor you choose.
+- [Metamask](https://metamask.io/) - For managing our wallet.
+- [Hardhat](https://hardhat.org/) - For compiling and deploying the smart contract
+- [Alfajores Testnet Account](https://developers.celo.org/3-simple-steps-to-connect-your-metamask-wallet-to-celo-732d4a139587) - Required to connect to the dApp and make test transactions.
+- [Node.js](https://nodejs.org/en/) - From V12. or higher
+
+## Witnet and Celo
+
+Witnet is a decentralized oracle network that allows smart contracts to securely and reliably access external data sources, off-chain computations, and other off-chain resources. Celo, on the other hand, is a blockchain platform that aims to make financial tools and services more accessible to anyone with a mobile phone.
+
+The two platforms are not directly related, but they can be used together to build decentralized applications that require access to external data. For example, a developer building a decentralized prediction market on Celo might use Witnet as the oracle network to retrieve external data such as sports scores or stock prices. In our case, we are about to use Witnet and Celo blockchain to get price feeds of some popular blockchain networks.
+
+## Smart Contract
+
+The PriceFeeds smart contract retrieves price feeds data from the Witnet decentralized oracle network.
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.11;
+
+import "witnet-solidity-bridge/contracts/interfaces/IWitnetPriceRouter.sol";
+import "witnet-solidity-bridge/contracts/interfaces/IWitnetPriceFeed.sol";
+
+contract PriceFeeds {
+   IWitnetPriceRouter public immutable witnetPriceRouter;
+   mapping(bytes4 => IWitnetPriceFeed) public priceFeeds;
+   bytes4[] ID4;
+   uint256 public lastUpdateTime;
+
+   constructor(IWitnetPriceRouter _router, bytes4[] memory _ID4) {
+       witnetPriceRouter = _router;
+       ID4 = _ID4;
+       lastUpdateTime = block.timestamp;
+       updatePriceFeeds();
+   }
+
+   function updatePriceFeeds() public {
```

Comment:
can you check out your code base, its kinda buggy, I mean the "updatePriceFeeds" function specifically, theres an error while compiling, can you look into it

Reply:
> can you check out your code base, its kinda buggy, I mean the "updatePriceFeeds" function specifically, theres an error while compiling, can you look into it

Recompiled from Remix and from the project, both are compiling. Did you create `.env` file and your REACT_APP_PRIVATE_KEY value?